action/action.go

```diff
@@ -17,15 +21,160 @@ import (
+	if err != nil {
+		return driver.OperationResult{}, claim.Result{}, err
+	}
+
+	err = OperationConfigs(opCfgs).ApplyConfig(op)
+	if err != nil {
+		return driver.OperationResult{}, claim.Result{}, err
+	}
+
+	var opErr *multierror.Error
+	opResult, err := a.Driver.Run(op)
+	if err != nil {
+		opErr = multierror.Append(opErr, err)
+	}
+
+	err = opResult.SetDefaultOutputValues(*op)
+	if err != nil {
+		opErr = multierror.Append(opErr, err)
+	}
+
+	cr, err := buildClaimResult(c, opResult, opErr)
+	if err != nil {
+		opErr = multierror.Append(opErr, err)
+	}
+
+	// These are any errors from running the operation or processing the result,
+	// We don't return it as an error because at this point the bundle has been
+	// executed and we are returning results that should be persisted. We don't
+	// want someone checking if an error occurred then ignoring the other return
+	// values.
+	opResult.Error = opErr.ErrorOrNil()
+
+	return opResult, cr, nil
+}
+
+// SaveInitialClaim with the specified status. If not used, the caller is
+// responsible for persisting the claim.
+func (a Action) SaveInitialClaim(c claim.Claim, status string) error {
+	if a.Claims == nil {
+		return errors.New("the action claims provider is not set")
+	}
+
+	err := a.saveClaimWithStatus(c, status)
+	return errors.Wrap(err, "could not save the pending action's status, the bundle was not executed")
+}
+
+// SaveOperationResult saves the ClaimResult and Outputs. The caller is
+// responsible for having already persisted the claim itself, for example using
+// SaveInitialClaim.
+func (a Action) SaveOperationResult(opResult driver.OperationResult, c claim.Claim, r claim.Result) error {
+	if a.Claims == nil {
+		return errors.New("the action claims provider is not set")
+	}
+
+	// Keep accumulating errors from any error returned from the operation
+	// We must save the claim even when the op failed, but we want to report
+	// ALL errors back.
+	var bigerr *multierror.Error
+	bigerr = multierror.Append(bigerr, opResult.Error)
+
+	err := a.Claims.SaveResult(r)
+	if err != nil {
+		bigerr = multierror.Append(bigerr, err)
+	}
+
+	for outputName, outputValue := range opResult.Outputs {
+		if !a.shouldSaveOutput(outputName) {
+			continue
+		}
+
+		output := claim.Output{
+			Claim:  c,
+			Result: r,
+			Name:   outputName,
+			Value:  []byte(outputValue),
+		}
+		err = a.Claims.SaveOutput(output)
+		if err != nil {
+			bigerr = multierror.Append(bigerr, err)
+		}
+	}
+
+	return bigerr.ErrorOrNil()
+}
+
+func (a Action) shouldSaveOutput(name string) bool {
+	if a.SaveAllOutputs {
+		return true
+	}
+	for _, output := range a.SaveOutputs {
+		if name == output {
+			return true
+		}
+	}
+	return false
+}
+
+// saveClaimWithStatus saves a claim and a result with the specified status.
+func (a Action) saveClaimWithStatus(c claim.Claim, status string) error {
+	r, err := c.NewResult(status)
+	if err != nil {
+		return err
+	}
+
+	err = r.Validate()
```

Comment:
I thought about that for bit but I would recommend finding a way to eliminate the error from the return parameters instead of doubling down on it.

I think we need to find a solution for creating ULIDs that doesn't have errors as a possibility. We rely on these documents to store critical data and logs, and not knowing if we even can store them in some cases is a problem.

I have ugly hacks in my back pocket to work around this*, but really we shouldn't be having to workaround it in the first place 😝  and should be able to safely generate a ULID and persist data about the state of the system.

\* essentially generate the ULIDs we need before we attempt the operation and munge the signature of these constructors to take the ULID up-front

Reply:
I have created #216 to track